File: app/src/main/java/com/ternaryop/photoshelf/db/TumblrPostCacheDAO.java
package com.ternaryop.photoshelf.db;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.provider.BaseColumns;
import android.support.annotation.NonNull;

import com.ternaryop.tumblr.TumblrPost;

public class TumblrPostCacheDAO extends AbsDAO<TumblrPostCache> implements BaseColumns {
    public static final String TABLE_NAME = "tumblr_post_cache";

    public static final String BLOG_NAME = "blog_name";
    public static final String CACHE_TYPE = "cache_type";
    public static final String TIMESTAMP = "post_timestamp";
    public static final String POST = "post_object";

    public static final String[] COLUMNS = new String[] { _ID, BLOG_NAME, CACHE_TYPE, TIMESTAMP, POST };

    /**
     * Constructor is accessible only from package
     */
    TumblrPostCacheDAO(SQLiteOpenHelper dbHelper) {
        super(dbHelper);
    }

    protected void onCreate(SQLiteDatabase db) {
        String sql = "CREATE TABLE tumblr_post_cache(" +
                "_id                    TEXT NOT NULL," +
                "blog_name              TEXT NOT NULL," +
                "cache_type             INTEGER NOT NULL," +
                "post_timestamp         INT UNSIGNED NOT NULL," +
                "post_object            BLOB NOT NULL," +
                "PRIMARY KEY(_id))";
        db.execSQL(sql);
        db.execSQL(String.format("CREATE INDEX %1$s_%2$s_%3$s_IDX ON %1$s(%2$s, %3$s);", TABLE_NAME, BLOG_NAME, CACHE_TYPE));
    }

    protected void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        if (newVersion == 2) {
            onCreate(db);
        }
    }

    public ContentValues getContentValues(TumblrPostCache post) {
        ContentValues v = new ContentValues();

        try {
            v.put(_ID,  post.getId());
            v.put(BLOG_NAME, post.getBlogName());
            v.put(CACHE_TYPE, post.getCacheType());
            v.put(TIMESTAMP, post.getTimestamp());
            v.put(POST, toBlob(post.getPost()));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        return v;
    }

    @Override
    public String getTableName() {
        return TABLE_NAME;
    }

    public int deleteItem(TumblrPost item) {
        SQLiteDatabase db = getDbHelper().getWritableDatabase();
        return db.delete(TABLE_NAME, _ID + "=? and " + BLOG_NAME + "=?", new String[] {item.getReblogKey(), item.getBlogName()});
    }

    public boolean updateItem(TumblrPost post, int cacheType) {
        final TumblrPostCache cache = new TumblrPostCache(post.getReblogKey(), post, cacheType);

        ContentValues v = getContentValues(cache);

        return getDbHelper().getWritableDatabase().update(TABLE_NAME, v, _ID + "=?", new String[] {cache.getId()}) == 1;
    }

    public void clearCache(int cacheType) {
        getDbHelper().getWritableDatabase().delete(TABLE_NAME, CACHE_TYPE + "=?", new String[] {String.valueOf(cacheType)});
    }

    public List<TumblrPost> read(String blogName, int cacheType) {
        SQLiteDatabase db = getDbHelper().getReadableDatabase();
        String sqlQuery = "SELECT " + POST
                + " FROM " + TABLE_NAME
                + " WHERE lower(" + BLOG_NAME +") = ?"
                + " AND " + CACHE_TYPE + " =?";

        ArrayList<TumblrPost> list = new ArrayList<>();
        try (Cursor c = db.rawQuery(sqlQuery, new String[]{blogName.toLowerCase(Locale.US), String.valueOf(cacheType)})) {
            while (c.moveToNext()) {
                list.add((TumblrPost) fromBlob(c.getBlob(0)));
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        return list;
    }

    public void write(Collection<TumblrPost> posts, int cacheType) {
        for (TumblrPost post : posts) {
            insert(new TumblrPostCache(post.getReblogKey(), post, cacheType));
        }
    }

    public void delete(List<TumblrPost> posts, int cacheType) {
        if (posts.isEmpty()) {
            return;
        }
        final Collection<String> postIds = getIds(posts);

        String whereClause = _ID + " in (" + inClauseParameters(postIds.size()) + ") and " + BLOG_NAME + "=? and " + CACHE_TYPE + "=?";
        String[] args = new String[postIds.size() + 2];
        postIds.toArray(args);
        args[args.length - 2] = posts.get(0).getBlogName();
        args[args.length - 1] = String.valueOf(cacheType);

        getDbHelper().getWritableDatabase().delete(TABLE_NAME, whereClause, args);
    }

    public long findMostRecentTimestamp(String blogName, int cacheType) {
        SQLiteDatabase db = getDbHelper().getReadableDatabase();
        String sqlQuery = "SELECT max(" + TIMESTAMP + ")"
                + " FROM " + TABLE_NAME
                + " WHERE lower(" + BLOG_NAME +") = ?"
                + " AND " + CACHE_TYPE + " =?";

        try (Cursor c = db.rawQuery(sqlQuery, new String[]{blogName.toLowerCase(Locale.US), String.valueOf(cacheType)})) {
            if (c.moveToNext()) {
                return c.getLong(0);
            }
        }
        return 0;
    }

    @NonNull
    private Collection<String> getIds(Collection<TumblrPost> posts) {
        HashSet<String> ids = new HashSet<>();
        for (TumblrPost post : posts) {
            ids.add(post.getReblogKey());
        }
        return ids;
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/db/TumblrPostCacheDAO.kt
@file:Suppress("MaxLineLength")
package com.ternaryop.photoshelf.db

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.provider.BaseColumns
import com.ternaryop.tumblr.TumblrPost
import java.io.IOException
import java.util.Locale

@Suppress("TooManyFunctions", "ObjectPropertyNaming")
class TumblrPostCacheDAO internal constructor(dbHelper: SQLiteOpenHelper) : AbsDAO<TumblrPostCache>(dbHelper), BaseColumns {
    override val tableName: String
        get() = TABLE_NAME

    override fun onCreate(db: SQLiteDatabase) {
        val sql = "CREATE TABLE tumblr_post_cache(" +
                "_id                    TEXT NOT NULL," +
                "blog_name              TEXT NOT NULL," +
                "cache_type             INTEGER NOT NULL," +
                "post_timestamp         INT UNSIGNED NOT NULL," +
                "post_object            BLOB NOT NULL," +
                "PRIMARY KEY(_id))"
        db.execSQL(sql)
        db.execSQL(String.format("CREATE INDEX %1\$s_%2\$s_%3\$s_IDX ON %1\$s(%2\$s, %3\$s);", TABLE_NAME, BLOG_NAME, CACHE_TYPE))
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (newVersion == 2) {
            onCreate(db)
        }
    }

    override fun getContentValues(pojo: TumblrPostCache): ContentValues {
        val v = ContentValues()

        try {
            v.put(BaseColumns._ID, pojo.id)
            v.put(BLOG_NAME, pojo.blogName)
            v.put(CACHE_TYPE, pojo.cacheType)
            v.put(TIMESTAMP, pojo.timestamp)
            v.put(POST, AbsDAO.toBlob(pojo.post!!))
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        return v
    }

    fun deleteItem(item: TumblrPost): Int = dbHelper.writableDatabase.delete(TABLE_NAME, BaseColumns._ID + "=? and " + BLOG_NAME + "=?", arrayOf(item.reblogKey, item.blogName))

    fun insertItem(post: TumblrPost, cacheType: Int): Long {
        return insert(TumblrPostCache(post.reblogKey, post, cacheType))
    }

    fun updateItem(post: TumblrPost, cacheType: Int): Boolean {
        val cache = TumblrPostCache(post.reblogKey, post, cacheType)

        val v = getContentValues(cache)

        return dbHelper.writableDatabase.update(TABLE_NAME, v, BaseColumns._ID + "=?", arrayOf(cache.id)) == 1
    }

    fun clearCache(cacheType: Int) {
        dbHelper.writableDatabase.delete(TABLE_NAME, "$CACHE_TYPE=?", arrayOf(cacheType.toString()))
    }

    fun read(blogName: String, cacheType: Int): List<TumblrPost> {
        val db = dbHelper.readableDatabase
        val sqlQuery = ("SELECT " + POST
                + " FROM " + TABLE_NAME
                + " WHERE lower(" + BLOG_NAME + ") = ?"
                + " AND " + CACHE_TYPE + " =?")

        val list = mutableListOf<TumblrPost>()
        try {
            db.rawQuery(sqlQuery, arrayOf(blogName.toLowerCase(Locale.US), cacheType.toString())).use { c ->
                while (c.moveToNext()) {
                    list.add(AbsDAO.fromBlob(c.getBlob(0)) as TumblrPost)
                }
            }
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

        return list
    }

    fun write(posts: Collection<TumblrPost>, cacheType: Int) {
        for (post in posts) {
            insert(TumblrPostCache(post.reblogKey, post, cacheType))
        }
    }

    fun delete(posts: List<TumblrPost>, cacheType: Int) {
        if (posts.isEmpty()) {
            return
        }
        delete(getIds(posts), cacheType, posts[0].blogName)
    }

    fun delete(postIds: Collection<String>, cacheType: Int, blogName: String) {
        if (postIds.isEmpty()) {
            return
        }
        val whereClause = BaseColumns._ID + " in (" + inClauseParameters(postIds.size) + ") and " + BLOG_NAME + "=? and " + CACHE_TYPE + "=?"
        val args = arrayOfNulls<String>(postIds.size + 2)

        var i = 0
        for (id in postIds) {
            args[i++] = id
        }
        args[args.size - 2] = blogName
        args[args.size - 1] = cacheType.toString()

        dbHelper.writableDatabase.delete(TABLE_NAME, whereClause, args)
    }

    fun findMostRecentTimestamp(blogName: String, cacheType: Int): Long {
        val db = dbHelper.readableDatabase
        val sqlQuery = ("SELECT max(" + TIMESTAMP + ")"
                + " FROM " + TABLE_NAME
                + " WHERE lower(" + BLOG_NAME + ") = ?"
                + " AND " + CACHE_TYPE + " =?")

        db.rawQuery(sqlQuery, arrayOf(blogName.toLowerCase(Locale.US), cacheType.toString())).use { c ->
            if (c.moveToNext()) {
                return c.getLong(0)
            }
        }
        return 0
    }

    private fun getIds(posts: Collection<TumblrPost>): Collection<String> {
        val ids = HashSet<String>()
        for (post in posts) {
            ids.add(post.reblogKey)
        }
        return ids
    }

    companion object {
        const val TABLE_NAME = "tumblr_post_cache"

        const val BLOG_NAME = "blog_name"
        const val CACHE_TYPE = "cache_type"
        const val TIMESTAMP = "post_timestamp"
        const val POST = "post_object"
    }
}